File: frontend/lib/model-registry.ts
export const MODEL_REGISTRY = {
  regression: {
    linear_regression: {
      name: "Linear Regression",
      description: "Simple linear model for continuous predictions",
      icon: "TrendingUp",
    },
    ridge: {
      name: "Ridge Regression",
      description: "Linear regression with L2 regularization",
      icon: "TrendingUp",
    },
    lasso: {
      name: "Lasso Regression",
      description: "Linear regression with L1 regularization",
      icon: "TrendingUp",
    },
    svr: {
      name: "Support Vector Regression",
      description: "Non-linear regression using support vectors",
      icon: "TrendingUp",
    },
    random_forest: {
      name: "Random Forest",
      description: "Ensemble of decision trees for robust predictions",
      icon: "Trees",
    },
    gradient_boosting: {
      name: "Gradient Boosting",
      description: "Sequential ensemble learning method",
      icon: "TrendingUp",
    },
    decision_tree: {
      name: "Decision Tree",
      description: "Tree-based model for interpretable predictions",
      icon: "GitBranch",
    },
    knn: {
      name: "K-Nearest Neighbors",
      description: "Instance-based learning algorithm",
      icon: "Network",
    },
  },
  classification: {
    logistic_regression: {
      name: "Logistic Regression",
      description: "Linear model for binary/multiclass classification",
      icon: "BarChart3",
    },
    svc: {
      name: "Support Vector Classifier",
      description: "Non-linear classification using support vectors",
      icon: "BarChart3",
    },
    random_forest: {
      name: "Random Forest",
      description: "Ensemble of decision trees for classification",
      icon: "Trees",
    },
    gradient_boosting: {
      name: "Gradient Boosting",
      description: "Sequential ensemble for classification",
      icon: "BarChart3",
    },
    decision_tree: {
      name: "Decision Tree",
      description: "Tree-based classifier",
      icon: "GitBranch",
    },
    knn: {
      name: "K-Nearest Neighbors",
      description: "Instance-based classification",
      icon: "Network",
    },
  },
}

export type ProblemType = "regression" | "classification" | "auto"
export type ModelKey = keyof typeof MODEL_REGISTRY.regression | keyof typeof MODEL_REGISTRY.classification

File: backend/app/core/model_registry.py
# app/core/model_registry.py
from sklearn.linear_model import LinearRegression, Ridge, Lasso, LogisticRegression
from sklearn.svm import SVR, SVC
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier, GradientBoostingRegressor, GradientBoostingClassifier
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier

MODEL_REGISTRY = {
    "regression": {
        "linear_regression": LinearRegression,
        "ridge": Ridge,
        "lasso": Lasso,
        "svr": SVR,
        "random_forest": RandomForestRegressor,
        "gradient_boosting": GradientBoostingRegressor,
        "decision_tree": DecisionTreeRegressor,
        "knn": KNeighborsRegressor,
    },
    "classification": {
        "logistic_regression": LogisticRegression,
        "svc": SVC,
        "random_forest": RandomForestClassifier,
        "gradient_boosting": GradientBoostingClassifier,
        "decision_tree": DecisionTreeClassifier,
        "knn": KNeighborsClassifier,
    }
}

def get_model(problem_type: str, model_type: str):
    try:
        return MODEL_REGISTRY[problem_type][model_type]
    except KeyError:
        raise ValueError(f"Unknown model type '{model_type}' for problem '{problem_type}'.")
